library/x509write_csr.c

```diff
@@ -122,7 +130,8 @@ int mbedtls_x509write_csr_set_ns_cert_type( mbedtls_x509write_csr *ctx,
     return( 0 );
 }
 
-int mbedtls_x509write_csr_der( mbedtls_x509write_csr *ctx, unsigned char *buf, size_t size,
+static int mbedtls_x509write_csr_der_internal( mbedtls_x509write_csr *ctx, unsigned char *buf, size_t size,
```

Comment:
Why do you need to create this internal function?
Why not keep the original implementation, except instead of having the `tmp_buf` on the stack, allocate it on the heap, with size `size` which is a already given as an input parameter for the size of `buf`.
`tmp_buf` cannot exceed the size of `buf` anyway, so I don't see an issue, except maybe a memory allocation error.
This will make the `tmp_buf` size dynamic, and we can generate large csr as well.
We may consider removing the `tmp_buf` altogether, to make the code more efficient, but this requires further investigation and thinking.

Reply:
I created the internal function for making the code cleaner. The alternative was to make the memory allocation directly in mbedtls_x509write_csr_der but then I would have to check the return code for all the MBEDTLS_ASN1_CHK_ADD calls and free the memory on error. Is my interpretation wrong?